### PR TITLE
C22 -Phoenix -- Beenish Ali

### PR DESCRIPTION
**Pull request by beenishali693**

### CODE REVIEW

File: src/components/ChatEntry.jsx [hunk 1]
```diff
@@ -1,20 +1,43 @@
 import './ChatEntry.css';
+import PropTypes from 'prop-types';
+import TimeStamp from './TimeStamp';
+
+const ChatEntry = ({sender, body, timeStamp, id, liked, onLiked}) => {
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:5]: 👍 I like using destructured props to make it more visibly clear in the function definition itself what props we're expecting to receive. We do need to remember that these are all passed in as a single object (the one we usually call `props`) and it can cause problems if we forget to include the destructuring syntax (it's easy to forget and list the props as multiple separate params) I really prefer the glanceability.

File: src/components/ChatEntry.jsx [hunk 2]
```diff
@@ -1,20 +1,43 @@
...
-        <button className="like">🤍</button>
-      </section>
-    </div>
+    <>
+      <div className={localOrRemote}>
+        <h2 className="entry-name">{sender}</h2>
+        <section className="entry-bubble">
+          <p>{body}</p>
+          <p className="entry-time">
+            <TimeStamp time = {timeStamp} />
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:21]: Nice use of the supplied `TimeStamp`. All we need to do is pass in the `timeStamp` string from the message data and it takes care of the rest. All we had to do was confirm the name and type of the prop it was expecting (which we could do through its PropTypes) and we're all set!

Nit: no spaces around attribute `=`

File: src/components/ChatEntry.jsx [hunk 3]
```diff
@@ -1,20 +1,43 @@
...
 };
 
 ChatEntry.propTypes = {
-  // Fill with correct proptypes
+  sender: PropTypes.string.isRequired,
+  body: PropTypes.string.isRequired,
+  timeStamp: PropTypes.string.isRequired,
+  id: PropTypes.number.isRequired,
+  liked: PropTypes.bool.isRequired,
+  onLiked: PropTypes.func,
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:40]: 👍 The `id`, `sender`, `body`, `timeStamp`, and `liked` props are always passed (they're defined explicitly in the data and also provided in the test) so we can (and should) mark them `isRequired`, as you did. 

The remaining props were up to you, and the tests don't know about them. As a result, using `isRequired` causes a warning when running any tests that only pass the known props.

To properly mark any other props `isRequired`, we would also need to update the tests to include at least dummy values (such as an empty callback `() => {}` for the like handler) to make the proptype checking happy.

Alternatively, for any props that we leave not required, we should also have logic in our component to not try to use the value if it's undefined.

File: src/components/ChatLog.jsx [hunk 4]
```diff
@@ -0,0 +1,39 @@
+import './ChatLog.css';
+import ChatEntry from './ChatEntry';
+import PropTypes from 'prop-types';
+
+const ChatLog = ({entries, onLiked}) => {
+  const chatEntryComponents = entries.map((chatEntry) => {
+    return (
+      <ChatEntry
+        key={chatEntry.id}
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:9]: 👍 The `key` attribute is important for React to be able to detect certain kinds of data changes in an efficient manner. We're also using the `id` for our own `id` prop, so it might feel redundant to pass both, but one is for our logic and one is for React internals (we can't safely access the `key` value in any meaningful way).

File: src/components/ChatLog.jsx [hunk 5]
```diff
@@ -0,0 +1,39 @@
+import './ChatLog.css';
+import ChatEntry from './ChatEntry';
+import PropTypes from 'prop-types';
+
+const ChatLog = ({entries, onLiked}) => {
+  const chatEntryComponents = entries.map((chatEntry) => {
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:6]: Nice use of `map` to convert from the message data into `ChatEntry` components. We can perform this mapping storing the result into a variable we use in the JSX result as you did here (components are functions, so we can run JS code as usual before we reach the `return`, and even sometimes have multiple `return` statements with different JSX), we could make a helper function that we call as part of the `return`, or this expression itself could be part of the `return` JSX, which I often like since it helps me see the overall structure of the component, though can make debugging a little more tricky. But any of those approaches will work fine.

File: src/App.jsx [hunk 6]
```diff
@@ -1,16 +1,48 @@
 import './App.css';
+import Data from './data/messages.json';
```
Comment (anselrognlie) [inline src/App.jsx:2]: Prefer to use a lower camelcase symbol here. The import name becomes the variable name used in this file to refer to the data, which here is a list of data, not a class (yes React uses capital letters even though they're functions, but conceptually, they're like a class). An ALL_CAPS name would also be OK.

File: src/App.jsx [hunk 7]
```diff
@@ -1,16 +1,48 @@
...
+import ChatLog from './components/ChatLog';
+import { useState } from 'react';
+
+
+function App () {
+  const [messageData, setMessageData] = useState(Data);
+
+  const senders = [...new Set(Data.map(message => message.sender))];
+
+  const handleLiked = (id) => {
```
Comment (anselrognlie) [inline src/App.jsx:12]: 👍 Since our state is defined here, we also need to define our mutating function here so that it can "see" the setter function. All we need to receive is the `id` of the message to toggle, which allows us to locate the message to update, and calculate the next state value. We can pass this all the way down to our `ChatEntry` which handles the click event, passing us the `id` of the message that was clicked.

File: src/App.jsx [hunk 8]
```diff
@@ -1,16 +1,48 @@
...
+    setMessageData(messageData => messageData.map(message => {
```
Comment (anselrognlie) [inline src/App.jsx:13]: Nice use of the callback setter style. In this application, it doesn't really matter whether we use the callback style or the value style, but it's good practice to get in the habit of using the callback style.

Comment (anselrognlie) [reply src/App.jsx:13]: Nice use of map here to both handle making a new list so that React sees the message data has changed, and make new data for the clicked message with its like status toggled.

File: src/App.jsx [hunk 9]
```diff
@@ -1,16 +1,48 @@
...
+      if (message.id === id) {
+        return {...message, liked: !message.liked};
```
Comment (anselrognlie) [inline src/App.jsx:15]: We showed this approach in class, but technically, we're mixing a few responsibilities here. rather than this function needing to know _how_ to change the liked status itself, we could move this update logic to a helper function. This would better mirror how we eventually update records when there's an API call involved.

In this project, our messages are very simple objects, but if we had more involved operations, it could be worthwhile to create an actual class with methods to work with them, or at least have a set of dedicated helper functions to centralize any such mutation logic.

File: src/App.jsx [hunk 10]
```diff
@@ -1,16 +1,48 @@
...
+    let total = 0;
+    for (const message of messageData) {
+      if (message.liked) {
+        total += 1;
+      }
+    };
+    return total;
+  };
+
+  const totalLiked = calcTotalLiked(messageData);
```
Comment (anselrognlie) [inline src/App.jsx:32]: Nice job determining the total likes based on the like data of each message. We don't need an additional piece of state to track this, since it can be derived from the existing state we are tracking.

Explicitly totalling the count is perfectly fine, but many JS programmers would use `reduce` to achieve this.
```js
  const calcTotalLiked = (messageData) => {
    return messageData.reduce((acc, message) => {
      return message.liked ? acc + 1 : acc;
    }, 0);
  };
```
The first few times we work with `reduce`, it can be challenging to understand, but it's a tool that gets used commonly enough that it's worth practicing.

File: src/components/ChatEntry.jsx [hunk 11]
```diff
@@ -1,20 +1,43 @@
...
+import PropTypes from 'prop-types';
+import TimeStamp from './TimeStamp';
+
+const ChatEntry = ({sender, body, timeStamp, id, liked, onLiked}) => {
+  const heart = liked ? '❤️' :'🤍' ;
+
+  const localOrRemote = sender === 'Vladimir' ? 'chat-entry local' : 'chat-entry remote';
+
+  const onLikeClicked = () => {
+    onLiked(id);
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:11]: 👍 Passing the `id` of this message lets the logic defined up in the `App` find the message to update in its data.

File: src/components/ChatEntry.jsx [hunk 12]
```diff
@@ -1,20 +1,43 @@
 import './ChatEntry.css';
+import PropTypes from 'prop-types';
+import TimeStamp from './TimeStamp';
+
+const ChatEntry = ({sender, body, timeStamp, id, liked, onLiked}) => {
+  const heart = liked ? '❤️' :'🤍' ;
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:6]: 👍 We can figure out which emoji to use for the liked status based on the `liked` prop without creating any additional state.

File: src/components/ChatEntry.jsx [hunk 13]
```diff
@@ -1,20 +1,43 @@
...
+      <div className={localOrRemote}>
+        <h2 className="entry-name">{sender}</h2>
+        <section className="entry-bubble">
+          <p>{body}</p>
+          <p className="entry-time">
+            <TimeStamp time = {timeStamp} />
+          </p>
+          <button
+            className="like"
+            onClick={onLikeClicked}>
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:25]: 👍 We need a wrapper of some kind rather than calling the received callback through props, since our callback function is expecting a message id as its parameter. If we tried to use it directly as the click event handler, React would end up passing it a clink event, since any function registered as an event handler will always be given the event detail information as its argument.

File: src/components/ChatEntry.jsx [hunk 14]
```diff
@@ -1,20 +1,43 @@
 import './ChatEntry.css';
+import PropTypes from 'prop-types';
+import TimeStamp from './TimeStamp';
+
+const ChatEntry = ({sender, body, timeStamp, id, liked, onLiked}) => {
+  const heart = liked ? '❤️' :'🤍' ;
+
+  const localOrRemote = sender === 'Vladimir' ? 'chat-entry local' : 'chat-entry remote';
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:8]: Nice logic to decide whether to treat a message as local or remote. How could we generalize this so that it didn't need to look explicitly for Vladimir? This project only passes in a single conversation, but ideally, our components should work in other situations.

In the general case, does the `ChatEntry` itself have enough information as it is to "know" which messages are local and which are remote?

File: src/App.jsx [hunk 15]
```diff
@@ -1,16 +1,48 @@
 import './App.css';
+import Data from './data/messages.json';
+import ChatLog from './components/ChatLog';
+import { useState } from 'react';
+
+
+function App () {
+  const [messageData, setMessageData] = useState(Data);
+
+  const senders = [...new Set(Data.map(message => message.sender))];
```
Comment (anselrognlie) [inline src/App.jsx:10]: Nice use of `map` to extract the sender information, then uniquing that with a `Set` to get the participants in the conversation. We can confirm with the docs that iterating over the `Set` contents will occur in insertion order, meaning the resulting list of senders, however many there are, will be in the order they participated in the conversation.

File: src/App.jsx [hunk 16]
```diff
@@ -1,16 +1,48 @@
...
-      </header>
-      <main>
-        {/* Wave 01: Render one ChatEntry component
-        Wave 02: Render ChatLog component */}
-      </main>
-    </div>
+    <>
+      <div id="App">
+        <header>
+          <h1>Chat between {senders[0]} and {senders[1]}</h1>
```
Comment (anselrognlie) [inline src/App.jsx:38]: For the conversation in the project, there are exactly two participants and both have sent messages, which is the only reason we can find them. What would happen for other conversation situations?
- no one has sent a message yet
- only one participant has sent a message (the local user? a remote user?)
- there are more than two participants in this conversation

Some of these cases might not really be workable given the limited data we're working with, but it's worth thinking about what this could look like for a more complete application.

**Review by anselrognlie — COMMENTED** [review]
Looks good! Please review my comments, and let me know if you have any questions. Nice job!
